File: app/omnuv/estate.cpp
#include "estate.h"

#include <QJsonDocument>
#include <QJsonObject>
#include <QNetworkReply>

OmnuvRead::OmnuvRead(const QString& name, Get get, QObject* parent)
    : QObject(parent), m_name(name), m_get(std::move(get))
{
}

void OmnuvRead::read(const QString& path)
{
    if (path != m_path) {
        forget();
        m_path = path;
    }

    // The sequence number is what drops a late answer: only the newest
    // question may land, and `forget()` moves it on without asking anything.
    const quint64 seq = ++m_seq;
    QNetworkReply* reply = m_get(path);
    connect(reply, &QNetworkReply::finished, this, [this, reply, seq]() {
        reply->deleteLater();
        if (seq != m_seq) {
            return;
        }

        const int code = reply->attribute(QNetworkRequest::HttpStatusCodeAttribute).toInt();
        const QByteArray body = reply->readAll();

        if (reply->error() == QNetworkReply::NoError) {
            QJsonParseError parsed;
            const QJsonDocument doc = QJsonDocument::fromJson(body, &parsed);
            if (parsed.error == QJsonParseError::NoError) {
                m_data = doc.toVariant();
                m_readAt = QDateTime::currentDateTime();
                settle(QStringLiteral("current"), QString());
                emit landed();
                return;
            }
            settle(m_data.isValid() ? QStringLiteral("stale") : QStringLiteral("unavailable"),
                   tr("Omnuv answered with something this version cannot read."));
            return;
        }

        // Core's own sentence first: it says `{"error": "..."}` and, on a
        // 500, carries the reference a person can quote. Ours only when there
        // is nothing of Core's to show, and then it says which of the two
        // things happened rather than a status number.
        QString why = QJsonDocument::fromJson(body).object().value(QStringLiteral("error")).toString().trimmed();
        if (why.isEmpty()) {
            why = code == 0 ? tr("Omnuv could not be reached.")
                            : tr("Omnuv could not answer this just now.");
        }
        settle(m_data.isValid() ? QStringLiteral("stale") : QStringLiteral("unavailable"), why);
    });
}

void OmnuvRead::forget()
{
    ++m_seq;
    m_path.clear();
    m_data = QVariant();
    m_readAt = QDateTime();
    m_problem.clear();
    if (m_state != QLatin1String("loading")) {
        m_state = QStringLiteral("loading");
    }
    emit changed();
}

void OmnuvRead::retry()
{
    if (!m_path.isEmpty()) {
        read(m_path);
    }
}

void OmnuvRead::settle(const QString& state, const QString& problem)
{
    // Logged on a change of state only. Every minute's identical answer is
    // not news, and a log full of it hides the one line that is.
    if (state != m_state || problem != m_problem) {
        if (problem.isEmpty()) {
            qInfo().noquote() << "omnuv: estate" << m_name << state;
        }
        else {
            qWarning().noquote() << "omnuv: estate" << m_name << state << "-" << problem;
        }
    }
    m_state = state;
    m_problem = problem;
    emit changed();
}

OmnuvEstate::OmnuvEstate(OmnuvRead::Get get, QObject* parent)
    : QObject(parent),
      m_members(new OmnuvRead(QStringLiteral("members"), get, this)),
      m_parked(new OmnuvRead(QStringLiteral("parked"), get, this)),
      m_networks(new OmnuvRead(QStringLiteral("networks"), get, this)),
      m_devices(new OmnuvRead(QStringLiteral("devices"), get, this)),
      m_endpoints(new OmnuvRead(QStringLiteral("endpoints"), get, this)),
      m_keys(new OmnuvRead(QStringLiteral("keys"), get, this)),
      m_usage(new OmnuvRead(QStringLiteral("usage"), get, this)),
      m_history(new OmnuvRead(QStringLiteral("history"), get, this))
{
    // Devices hang off the project's network, so they are asked about once
    // the network is known — and asked again only if it is a different one.
    connect(m_networks, &OmnuvRead::landed, this, [this]() {
        const QVariantList nets = m_networks->data().toList();
        if (nets.isEmpty()) {
            m_devices->forget();
            return;
        }
        const QString id = nets.first().toMap().value(QStringLiteral("id")).toString();
        m_devices->read(QStringLiteral("/v1/networks/%1/devices").arg(id));
    });
}

void OmnuvEstate::setIdentity(const QString& organizationName, const QString& role)
{
    const bool owner = role == QLatin1String("owner");
    if (organizationName == m_organizationName && owner == m_owner) {
        return;
    }
    m_organizationName = organizationName;
    m_owner = owner;
    emit identityChanged();
}

void OmnuvEstate::setProject(const QString& projectId)
{
    if (projectId == m_project) {
        return;
    }
    m_project = projectId;
    for (OmnuvRead* r : { m_parked, m_networks, m_devices, m_endpoints, m_keys, m_usage, m_history }) {
        r->forget();
    }
    m_ticks = 0;
}

void OmnuvEstate::refresh(bool everything)
{
    if (everything) {
        m_ticks = 0;
    }
    if (m_ticks++ % 4 == 0) {
        readAll();
    }
    else {
        readFast();
    }
}

void OmnuvEstate::clear()
{
    m_project.clear();
    m_ticks = 0;
    for (OmnuvRead* r : { m_members, m_parked, m_networks, m_devices, m_endpoints, m_keys, m_usage, m_history }) {
        r->forget();
    }
    setIdentity(QString(), QString());
}

void OmnuvEstate::retryAll()
{
    for (OmnuvRead* r : { m_members, m_parked, m_networks, m_devices, m_endpoints, m_keys, m_usage, m_history }) {
        r->retry();
    }
}

void OmnuvEstate::readAll()
{
    readFast();
    readSlow();
}

void OmnuvEstate::readFast()
{
    // Nothing project-scoped is asked before the project is known: an
    // unscoped read answers for Core's default, which may not be the project
    // on screen.
    if (m_project.isEmpty()) {
        return;
    }
    const QString scope = QStringLiteral("?project=") + m_project;
    m_parked->read(QStringLiteral("/v1/parked") + scope);
    // Three, because that is what the window shows. Paged by `limit` alone:
    // Core computes the chain over the rows it returns, and a `from` would
    // leave the oldest of them without a predecessor.
    m_history->read(QStringLiteral("/v1/projects/%1/history?limit=3").arg(m_project));
}

void OmnuvEstate::readSlow()
{
    m_members->read(QStringLiteral("/v1/org/members"));
    if (m_project.isEmpty()) {
        return;
    }
    const QString scope = QStringLiteral("?project=") + m_project;
    // Devices are read again when this answer lands; see the constructor.
    m_networks->read(QStringLiteral("/v1/networks") + scope);
    m_endpoints->read(QStringLiteral("/v1/endpoints") + scope);
    m_keys->read(QStringLiteral("/v1/projects/%1/api-keys").arg(m_project));
    m_usage->read(QStringLiteral("/v1/usage?days=30&project=") + m_project);
}

File: app/omnuv/estate.js
// Omnuv: the estate's words — what the window says about a project, built
// from Core's fields and nothing else.
//
// **Ported from the web console, not reinvented.** `HistoryBand.svelte`,
// `format.ts` and the estate page say these sentences first; a change that
// reads one way in the browser and another in the tray is the product telling
// a person two things about one event. When the console's wording moves, this
// file moves with it.
//
// A library, so every file that imports it shares one copy, and pure, so
// `test/estate_test.js` can run it under node without a Qt.
.pragma library

// A buyer's vocabulary for each table the map records. The subject carries its
// own article so "an SSH key" needs no vowel test.
//
// `instances.create` is *requested*, because the epoch is minted when the
// marketplace agreed, before anything booted. Every `update` is *changed*:
// the history endpoint does not return before and after, and naming what
// changed from the intent alone would be asserting an observation nobody made.
var WORDS = {
    instances:          { subject: "a machine",         create: "requested", update: "changed", delete: "deleted" },
    networks:           { subject: "a private network", create: "created",   update: "changed", delete: "removed" },
    network_addresses:  { subject: "an address",        create: "assigned",  update: "changed", delete: "released" },
    endpoints:          { subject: "a public endpoint", create: "published", update: "changed", delete: "withdrawn" },
    recipe_deployments: { subject: "a deployment",      create: "deployed",  update: "changed", delete: "removed" },
    ssh_keys:           { subject: "an SSH key",        create: "added",     update: "changed", delete: "removed" },
    api_keys:           { subject: "an API key",        create: "created",   update: "changed", delete: "revoked" },
    devices:            { subject: "a device",          create: "enrolled",  update: "changed", delete: "removed" }
}

var UUID = /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i

// The name in front of the sentence, or nothing. A uuid is never printed: it
// is not a name and a buyer cannot act on it. `?` is the map's sentinel for a
// row with neither an id nor an address, and is as unprintable.
function nameOf(c) {
    var named = String(c.resource_name || "").trim()
    if (named) {
        return named
    }
    // An address's id is the address itself, which a buyer does read.
    var v = String(c.resource_id || "").trim()
    if (!v || v === "?" || UUID.test(v)) {
        return ""
    }
    return v
}

// Core's reasons describe the change in the third person ("the buyer created
// an api key"). The actor is already on the row, so it comes off; what is left
// is dropped when it only repeats the verb. A reason that adds something —
// "no free RTX 3090, nothing is charged" — survives.
function reasonClause(reason, verb) {
    var rest = String(reason || "").replace(/^(the buyer|the marketplace|omnuv|a member)\s+/i, "").trim()
    if (!rest || rest.toLowerCase().indexOf(verb.toLowerCase()) === 0) {
        return ""
    }
    return rest
}

function sentence(c) {
    // An unmapped table is still shown, in the least wrong words available:
    // silence about a change is worse than a clumsy sentence about it.
    var w = WORDS[c.resource] || {
        subject: "a " + String(c.resource).replace(/_/g, " ").replace(/s$/, ""),
        create: "created", update: "changed", delete: "removed"
    }
    var verb = w[c.intent] || "changed"
    var name = nameOf(c)
    var head = name ? name + " " + verb : w.subject + " was " + verb
    if (!name) {
        head = head.charAt(0).toUpperCase() + head.slice(1)
    }
    var why = reasonClause(c.reason, verb)
    return why ? head + " — " + why : head
}

// "just now", "4m ago", "in 7d". `now` is milliseconds, passed in so a view
// can tick it and a test can fix it.
function ago(iso, now) {
    if (!iso) {
        return "never"
    }
    var at = new Date(iso).getTime()
    if (isNaN(at)) {
        return ""
    }
    var diff = now - at
    var ahead = diff < 0
    var mins = Math.round(Math.abs(diff) / 60000)
    function wrap(s) { return ahead ? "in " + s : s + " ago" }
    if (mins < 1) {
        return ahead ? "now" : "just now"
    }
    if (mins < 60) {
        return wrap(mins + "m")
    }
    var hours = Math.round(mins / 60)
    if (hours < 24) {
        return wrap(hours + "h")
    }
    return wrap(Math.round(hours / 24) + "d")
}

// Two decimals for anything a person would pay in cash, and enough to be
// non-zero below a cent — a €0.0013 inference is not €0.00.
function money(value, currency) {
    var n = Number(value)
    if (isNaN(n)) {
        return "—"
    }
    var c = currency || "EUR"
    var symbol = c === "EUR" ? "€" : (c === "USD" ? "$" : c + " ")
    if (n !== 0 && Math.abs(n) < 0.01) {
        return symbol + n.toFixed(6).replace(/0+$/, "")
    }
    return symbol + n.toFixed(2)
}

function bytes(n) {
    if (n < 1024) {
        return n + " B"
    }
    var units = ["KB", "MB", "GB", "TB"]
    var v = n / 1024
    var u = 0
    while (v >= 1024 && u < units.length - 1) {
        v /= 1024
        u++
    }
    return v.toFixed(v < 10 ? 1 : 0) + " " + units[u]
}

// Thirty points, one per day, oldest first. Core sends `by_day` sparsely — a
// row only for a day something was spent — so a day it did not mention is a
// day nothing was spent, not a gap to draw across. Days are matched in UTC,
// as Core truncates them.
function series(usage, now) {
    if (!usage || !usage.by_day) {
        return []
    }
    var byDay = {}
    for (var i = 0; i < usage.by_day.length; i++) {
        byDay[usage.by_day[i].day] = Number(usage.by_day[i].cost)
    }
    var out = []
    for (var d = 29; d >= 0; d--) {
        var day = new Date(now - d * 86400000).toISOString().slice(0, 10)
        out.push(byDay[day] || 0)
    }
    return out
}

// The pulse row. A counter whose source could not be read is **left out**,
// never shown as 0: a failed read is not an empty estate.
function pulse(counts, machinesLoaded, parked) {
    var out = []
    if (machinesLoaded) {
        out.push({ label: "running", n: counts["Running"] || 0, tone: "running" })
        out.push({ label: "starting", n: counts["Starting"] || 0, tone: "starting" })
    }
    if (parked) {
        out.push({ label: "waiting", n: waiting(parked).length, tone: "waiting" })
    }
    if (machinesLoaded) {
        out.push({ label: "stopped", n: counts["Stopped"] || 0, tone: "stopped" })
    }
    return out
}

function waiting(parked) {
    var out = []
    for (var i = 0; i < (parked || []).length; i++) {
        if (parked[i].status === "WAITING") {
            out.push(parked[i])
        }
    }
    return out
}

// One banner for the band, not one per read: a stale read is a statement about
// the screen. Returns the reads that could not refresh.
function staleOf(reads) {
    var out = []
    for (var i = 0; i < reads.length; i++) {
        if (reads[i] && reads[i].state === "stale") {
            out.push(reads[i])
        }
    }
    return out
}
